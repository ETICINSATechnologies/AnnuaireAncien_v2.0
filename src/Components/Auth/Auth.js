class Auth {
    static connect(jwt){
        sessionStorage.setItem('token', jwt);
    }

    static disconnect(){
        sessionStorage.removeItem('token')
    }

    static isConnected(){
        return !!sessionStorage.getItem('token');
    }

    static getToken() {
        return sessionStorage.getItem('token')
    }

    static getPayload(){
        return sessionStorage.getItem('token');
    }

    static addCorrectButton(buttons){
        if (this.isConnected())
            buttons.push("disconnection");
        else
            buttons.push("connection");
        return buttons;
    }
}

export default Auth;
