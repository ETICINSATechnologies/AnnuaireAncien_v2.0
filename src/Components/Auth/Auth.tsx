class Auth {
    static connect(jwt: string) {
        sessionStorage.setItem('token', jwt);
    }

    static disconnect(): void {
        sessionStorage.removeItem('token')
    }

    static isConnected(): boolean {
        return !!sessionStorage.getItem('token');
    }

    static getToken(): string {
        return sessionStorage.getItem('token') as string;
    }

    static addCorrectButton(buttons: Array<string>) {
        if (this.isConnected())
            buttons.push("disconnection");
        else
            buttons.push("connection");
        return buttons;
    }
}

export default Auth;
