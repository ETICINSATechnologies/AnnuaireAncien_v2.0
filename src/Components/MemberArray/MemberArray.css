.MemberArray {
    display: grid;
}

.MemberArray table {
    font-family: Arial, Helvetica, sans-serif;
    border: 0.3vmin solid black;
    border-collapse: collapse;
    margin: auto;
    height: 100%;
}

.MemberArray .info {
    font-size: 2.2vw;
    margin: auto;
    padding-top: 10vw;
    text-align: center;
}

.MemberArray .searchResult{
    height: 100%;
    margin: 1vw 0 1vmin 0;
}

.MemberArray .data{
    font-size: 2.2vw;
    text-align: center;
    height : auto;
}

.MemberArray .pageInput{
    font-size: 1.1vmax;
    text-align: center;
    height: 1.2em;
    width: 1.5em;
}

.MemberArray input[type="number"]::-webkit-outer-spin-button, input[type="number"]::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

.MemberArray input[type="number"] {
    -moz-appearance: textfield;
}

.MemberArray caption {
    margin: 0 0 1vmin 0;
}

.MemberArray thead, tbody {
    table-layout: fixed;
    width: 100%;
}

.MemberArray thead {
    margin-bottom: 0.3vw;
    box-shadow: 0 0.2vw 0.3vw #888888;
}

.MemberArray tbody {
    overflow-y: auto;
    overflow-x: hidden;
}

.MemberArray td, th {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 1vmin;
    border-bottom: 0.2vmin solid black;
    border-left: 0.2vmin solid black;
}

.MemberArray td:first-child, th:first-child{
    border-left: 0;
}

.MemberArray tr:last-child td {
    border-bottom: 0;
}

.MemberArray td {
    cursor: pointer;
}

.MemberArray tr {
    background-color: white;
}

.MemberArray tr:nth-child(even) {
    background-color: #f2f2f2;
}

.MemberArray tr:hover {
    background-color: #ddd;
}

.MemberArray th {
    color: white;
    background-color: var(--component-bg-color);
    text-align: left;
}

.MemberArray .pagination {
    position: absolute;
    left: 0;
    right: 0;
    margin-left: auto;
    margin-right: auto;
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    font-family: Arial, Helvetica, sans-serif;
    justify-items: center;
    align-items: center;
    bottom: 3vh;
}

@media (orientation: landscape) {
    .MemberArray .pagination {
        width: 35vw;
    }
}

@media (orientation: portrait){
    .MemberArray .pagination {
        width: 80%;
    }
}

.MemberArray .pagination > input{
    grid-row: 1;
    text-align: center;
    vertical-align: middle;
    width: 10vmax;
    font-size: 1.1vmax;
    border: 0.2vmin solid black;
    height: 4.5vh;
}

.MemberArray .pagination p{
    grid-row:1;
    text-align: center;
    font-size: 1.1vmax;
    width: auto;
}

.MemberArray .deleteContainer .deletePicture {
    margin: auto;
    width: 2.7vw;
    max-width: 45px;
}

.MemberArray .Modal .content {
    font-size: 2.5vmin;
    padding: 2vh 2vw;
    display : grid;
    grid-template-rows: repeat(2,1fr);
    text-align: center;
    height: auto;
    width: auto;
}

.MemberArray .Modal .content .buttons{
    display : grid;
    grid-template-columns: repeat(2,1fr);
    max-height: 45px;
    margin: auto;
}

.MemberArray .Modal .content .buttons button{
    margin: auto 3vmin;
    padding: 0rem 2rem;
    font-size: 2.5vmin;
    max-width: 400px;
    background-color: #fff;
    border-width:1px;
    border-color: black;
}


