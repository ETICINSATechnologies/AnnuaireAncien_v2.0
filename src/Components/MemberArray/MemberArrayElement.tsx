import React, {Component} from 'react';
import {MemberRead} from "../../Model/Member";

interface MemberArrayElementProps {
    key: number
    memberInfo: MemberRead

    selectMemberById(id: number): void
}


const MemberArrayElement = (props: MemberArrayElementProps) =>
    <tr onClick={() => props.selectMemberById((props.memberInfo as any).id)}>
        <td className="firstName">{props.memberInfo.firstName}</td>
        <td className="lastName">{props.memberInfo.lastName}</td>
        <td className="positionLabel">{props.memberInfo.latestPosition}</td>
        <td className="positionYear">{props.memberInfo.latestPosition}</td>
    </tr>
;


/*class MemberArrayElements extends Component {
    constructor(props) {
        super(props);
        const propertiesName = {
            id: null,
            lastName: 'nom',
            firstName: 'prénom',
            positionLabel: 'poste',
            positionYear: 'année'
        };
        this.state = {
            header: this.props.header || false,
            propertiesName: propertiesName,
            properties: this.getMemberProperties(propertiesName)
        };
    }

    getPositionLabel() {
        try {
            if (this.props.info.positions[0].label)
                return this.props.info.positions[0].label;
        } catch (e) {
            if (e instanceof TypeError) return null;
        }
    }

    getPositionYear() {
        try {
            if (this.props.info.positions) {
                if (this.props.info.positions[0].year)
                    return this.props.info.positions[0].year;
                return 2018;
            }
        } catch (e) {
            if (e instanceof TypeError) return 2018;
        }
    }

    getMemberProperties(propertiesName) {
        let newProperties = {};
        Object.keys(propertiesName).forEach((propertyName) => {
            if (this.props.info.hasOwnProperty(propertyName)) {
                newProperties[propertyName] = this.props.info[propertyName];
            }
        });
        newProperties.positionLabel = this.getPositionLabel();
        newProperties.positionYear = this.getPositionYear();

        return newProperties;
    }

    render() {
        return (
            <tr className={`MemberArrayElement ${this.state.properties.id}`}
                onClick={() => this.props.onClick(this.state.properties.id)}>
                {this.state.header
                    ?
                    <React.Fragment>
                        <th className='firstName'>{this.state.propertiesName.firstName}</th>
                        <th className='lastName'>{this.state.propertiesName.lastName}</th>
                        <th className='positionLabel'>{this.state.propertiesName.positionLabel}</th>
                        <th className='positionYear'>{this.state.propertiesName.positionYear}</th>
                    </React.Fragment>
                    :
                    <React.Fragment>
                        <td className='firstName'>{this.state.properties.firstName}</td>
                        <td className='lastName'>{this.state.properties.lastName}</td>
                        <td className='positionLabel'>{this.state.properties.positionLabel}</td>
                        <td className='positionYear'>{this.state.properties.positionYear}</td>
                    </React.Fragment>
                }
            </tr>
        );
    }
}*/

export default MemberArrayElement;
