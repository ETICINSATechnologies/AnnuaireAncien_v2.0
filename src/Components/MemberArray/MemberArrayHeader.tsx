import React, {Component} from 'react';

const MemberArrayHeader = () =>
    <tr>
        <th className="firstName">Prénom</th>
        <th className="lastName">Nom</th>
        <th className="positionLabel">Poste</th>
        <th className="positionYear">Année</th>
    </tr>
;

export default MemberArrayHeader;