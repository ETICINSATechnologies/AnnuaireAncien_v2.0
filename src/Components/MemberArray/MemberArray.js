import React, {Component} from 'react';
import './MemberArray.css';
import Auth from "../../Components/Auth/Auth";
import MemberDisplay from "../../Components/MembersDisplay/MemberDisplay";


class MemberArray extends Component {
    constructor(props) {
        super(props);
        this.state = {
            status: 'init', // 'noResult', 'result', 'pending
            mapping: {},
            info: {}
        };
        this.getMembers = this.getMembers.bind(this);
    }

    componentDidMount() {
        if (!this.props.search)
            this.getMembers();
    }

    getMembers(searchArray) {
        let url = 'api/v1/core/member';
        if (searchArray) {
            Object.keys(searchArray).forEach((key, i) => {
                if (i === 0) url += '?';
                else url += '&';
                url += key + '=' + searchArray[key];
            });
        }
        if (this.props.board) url += '/board/latest';
        this.setState({
            status: 'pending'
        });

        fetch(url, {
            headers: {
                Authorization: Auth.getToken()
            }
        })
            .then(res => res.json())
            .then((result) => {
                if (result && result.content.length > 0) {
                    let mapping = this.mapMembers(result);
                    this.setState({
                        status: 'result',
                        info: result,
                        mapping: mapping
                    })
                }
                else {
                    this.setState({
                        status: 'noResult'
                    })
                }
            })
    }

    /**
     Associate the id of a member with its position in the array 'state.info.content'
     */
    mapMembers(result) {
        let i = 0;
        let mapping = {};
        result.content.forEach((member) => {
            if (!(member.id in mapping))
                mapping[member.id] = i++;
        });

        return mapping
    }

    getMemberById(id) {
        if (id in this.state.mapping) {
            let position = this.state.mapping[id];
            return this.state.info.content[position];
        }
        return null
    }

    render() {
        let members = [];

        if (this.state.status === 'result' && this.state.info.hasOwnProperty('content')) {
            for (let i = 0; i < this.state.info.content.length; i++) {
                if (this.state.info.content[i])
                    members.push(
                        <MemberDisplay key={i} info={this.state.info.content[i]}
                                       onClick={(id) => this.props.onClick(this.getMemberById(id))}/>
                    )
            }
        }
        else {
            let message;
            if (this.state.status === 'init') message = "Vous pouvez effectuer une recherche sur les critères ci-contre";
            else if (this.state.status === 'noResult') message = "La recherche n'a donné aucun résultat";
            else if (this.state.status === 'pending') message = "Chargement...";

            return (
                <section className={`MemberArray ${this.props.className}`}>
                    <div className='info'>{message}</div>
                </section>
            );
        }

        return (
            <section className={`MemberArray ${this.props.className}`}>
                <table>
                    <caption>
                        Liste des membres {this.props.board ? 'du CA' : ''}
                    </caption>
                    <thead><MemberDisplay header info/></thead>
                    <tbody>{members}</tbody>
                </table>
            </section>
        );
    }
}

export default MemberArray;
