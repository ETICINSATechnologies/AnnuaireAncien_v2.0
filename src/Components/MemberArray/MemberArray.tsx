import React, {Component} from 'react';

import './MemberArray.css';
import Auth from "../../Components/Auth/Auth";

import {Member, MemberInterface} from "../../Model/Member";
import {SearchInterface} from "../../Model/Searchinterface";
import MemberArrayElement from "./MemberArrayElement";
import MemberArrayHeader from "./MemberArrayHeader";


interface MemberArrayProps {
    parameters: SearchInterface
    ref: string
    selectMember(member: Member): void
    getPage(pageNum : number):void
}

interface PageStatus{
    currentPage : number
    totalPages : number
}

interface MemberArrayState {
    status: statusType
    members: Member[]
    memberMapping: object
    page: PageStatus
    bufferInput: string
}

type statusType = 'init' | 'noResult' | 'result' | 'pending' | 'error';

class MemberArray extends Component<MemberArrayProps, MemberArrayState> {

    state = {
        status: 'init' as statusType,
        members: [] as Member[],
        memberMapping: {},
        page: {
            currentPage : 0,
            totalPages : 1
        },
        bufferInput: '1'
    };

    getPageSize = () => {
        if(window.matchMedia("(orientation:landscape)")){
            return 12;
        }else if(window.matchMedia("(orientation:portrait) and (max-width: 500px)")){
            return 15;
        }else{
            return 8;
        }
    }

    getMembers = (searchArray: SearchInterface, page : number) => {
        //manage pages
        let baseurl : string = 'member?pageSize='+this.getPageSize()+'&pageNumber='+page;

        // create the url with eventually the parameters
        let url = Object.keys(searchArray).reduce((url, property, i) => {
            let value = searchArray[property];
            if (value === '') return url;
            let delimiter = '&';
            return url + delimiter + property + '=' + value
        }, baseurl);

        this.setState({
            status: 'pending'
        });


        fetch(url, {
            headers: {
                Authorization: Auth.getToken()
            }
        })
            .then(res => {
                if (res.status === 200)
                    res.json()
                        .then((result) => {
                            if (result && result.content.length > 0) {
                                let memberMapping = {};
                                let members = result.content.map(
                                    (member: MemberInterface, index: string) => {
                                        (memberMapping as any)[member.id] = index;
                                        return new Member(member)
                                    });

                                let thisPage : PageStatus = {
                                    currentPage : result.meta.page,
                                    totalPages : result.meta.totalPages,
                                };

                                this.setState({
                                    status: 'result',
                                    members: members,
                                    memberMapping: memberMapping,
                                    page: thisPage,
                                    bufferInput: result.meta.page+1
                                })
                            }
                            else {
                                this.setState({
                                    status: 'noResult'
                                })
                            }
                        });
                else {
                    this.setState({
                        status: 'error'
                    })
                }
            })
    };

    selectMemberById = (id: number) => {
        this.props.selectMember(this.state.members[(this.state.memberMapping as any)[id]]);
    };

    selectPageByInput = (event: any) => {
        if(event.key === 'Enter') {
            this.props.getPage(event.target.value-1);
        }
    };

    inputChangeHandler = (event: any) => {
        this.setState({
            bufferInput: event.target.value
            }
        );
    }

    render() {
        let data;
        if (this.state.status === 'result') {
            const members = this.state.members.map((member) => {
                return <MemberArrayElement key={member.id} memberInfo={member.read()}
                                           selectMemberById={this.selectMemberById}/>;
            });
            data =
                    <div>
                        <table>
                            <caption>Liste des membres</caption>
                            <thead><MemberArrayHeader/></thead>
                            <tbody>{members}</tbody>
                        </table>
                    </div>
        }
        else {
            let message;
            if (this.state.status === 'init') message = "Vous pouvez effectuer une recherche sur les critères ci-contre";
            else if (this.state.status === 'noResult') message = "La recherche n'a donné aucun résultat";
            else if (this.state.status === 'error') message = "Une erreur est survenue lors de la recherche";
            else if (this.state.status === 'pending') message = "Chargement...";

            data = <div className='info'>{message}</div>;
        }


        return (
            <section className={`MemberArray members`}>
                <div className = 'searchResult'>
                    <div className='data'>
                        {data}
                    </div>
                    {this.state.page.totalPages===1? null :
                        <div className = 'pagination'>
                            <input disabled={this.state.page.currentPage === 0} type='button' className='pagePrevious'
                                   value='Page précédente' onClick={() => this.props.getPage(this.state.page.currentPage-1)}/>

                            <p className='pageIndicator'>
                                <label>Page </label>
                                <input type='number' className='pageInput' value={this.state.bufferInput} onKeyDown={this.selectPageByInput}
                                       onChange={this.inputChangeHandler}/>
                                <label> de {this.state.page.totalPages}</label>
                            </p>

                            <input disabled={(this.state.page.currentPage + 1) === this.state.page.totalPages}  type='button'
                                   className='pageNext' value='Page suivante'
                                   onClick={() => this.props.getPage(this.state.page.currentPage+1)}/>
                        </div>}
                </div>
            </section>
        )
    }
}

export default MemberArray;
