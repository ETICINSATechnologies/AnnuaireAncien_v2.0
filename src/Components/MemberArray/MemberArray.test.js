import React from 'react';
import MemberArray from './MemberArray';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Auth from '../../Components/Auth/Auth';

Enzyme.configure({adapter: new Adapter()});

describe( 'MemberArray component', () => {
    it('renders without crashing', () => {
            const tree = renderer.create(<MemberArray />).toJSON();
            expect(tree).toMatchSnapshot();
        }
    );

    it('MemberArray unit test', () => {
        expect(Enzyme.shallow(<MemberArray/>)
            .find('.MemberArray')
            .length
        ).toBe(1)
    });

});