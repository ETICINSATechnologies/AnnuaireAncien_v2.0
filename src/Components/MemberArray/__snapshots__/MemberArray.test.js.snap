// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberArray component renders without crashing 1`] = `
<section
  className="MemberArray undefined"
>
  <table>
    <caption>
      Liste des membres 
      
    </caption>
    <thead>
      <tr
        className="MemberDisplay undefined"
        onClick={[Function]}
      >
        <th
          className="firstName"
        >
          prénom
        </th>
        <th
          className="lastName"
        >
          nom
        </th>
        <th
          className="positionLabel"
        >
          poste
        </th>
        <th
          className="positionYear"
        >
          année
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</section>
`;
